confgenerator/apache.go

```diff
@@ -0,0 +1,167 @@
+				"Remove":    field,
+			},
+		})
+	}
+	// Generate the httpRequest structure.
+	c = append(c, fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Name":          "nest",
+			"Match":         tag,
+			"Operation":     "nest",
+			"Wildcard":      "http_request_*",
+			"Nest_under":    "logging.googleapis.com/http_request",
+			"Remove_prefix": "http_request_",
+		},
+	})
+	return c
+}
+
+type LoggingProcessorApacheError struct {
+	ConfigComponent `yaml:",inline"`
+}
+
+func (LoggingProcessorApacheError) Type() string {
+	return "apache_error"
+}
+
+func (p LoggingProcessorApacheError) Components(tag string, uid string) []fluentbit.Component {
+	c := LoggingProcessorParseRegex{
+		// Documentation: https://httpd.apache.org/docs/current/logs.html#errorlog
+		// Sample line 2.4: [Fri Sep 09 10:42:29.902022 2011] [core:error] [pid 35708:tid 4328636416] (13)Permission denied [client 72.15.99.187] File does not exist: /usr/local/apache2/htdocs/favicon.ico
+		// Sample line 2.2: [Fri Sep 09 10:42:29.902022 2011] [error] [pid 35708:tid 4328636416] [client 72.15.99.187] File does not exist: /usr/local/apache2/htdocs/favicon.ico
+		// TODO - Support time parsing for version 2.0 where smallest resolution is seconds
+		// Sample line 2.0: [Wed Oct 11 14:32:52 2000] [error] [client 127.0.0.1] client denied by server configuration: /export/home/live/ap/htdocs/test
+		Regex: `^\[(?<time>[^\]]+)\] \[(?:(?<module>\w+):)?(?<level>[\w\d]+)\](?: \[pid (?<pid>\d+)(?::tid (?<tid>[0-9]+))?\])?(?: (?<errorCode>[^\[]*))?(?: \[client (?<client>[^\]]*)\])? (?<message>.*)$`,
+		LoggingProcessorParseShared: LoggingProcessorParseShared{
+			TimeKey:    "time",
+			TimeFormat: "%a %b %d %H:%M:%S.%L %Y",
+			Types: map[string]string{
+				"pid": "integer",
+				"tid": "integer",
+			},
+		},
+	}.Components(tag, uid)
+	for _, l := range []struct{ level, severity string }{
+		{"emerg", "EMERGENCY"},
+		{"alert", "ALERT"},
+		{"crit", "CRITICAL"},
+		{"error", "ERROR"},
+		{"warn", "WARNING"},
+		{"notice", "NOTICE"},
+		{"info", "INFO"},
+		{"debug", "DEBUG"},
+		{"trace1", "DEBUG"},
+		{"trace2", "DEBUG"},
+		{"trace3", "DEBUG"},
+		{"trace4", "DEBUG"},
+		{"trace5", "DEBUG"},
+		{"trace6", "DEBUG"},
+		{"trace7", "DEBUG"},
+		{"trace8", "DEBUG"},
+	} {
+		c = append(c, fluentbit.Component{
+			Kind: "FILTER",
+			Config: map[string]string{
+				"Name":      "modify",
+				"Match":     tag,
+				"Condition": fmt.Sprintf("Key_Value_Equals level %s", l.level),
+				"Add":       fmt.Sprintf("logging.googleapis.com/severity %s", l.severity),
+			},
+		})
+	}
+	return c
+}
+
+type LoggingReceiverApacheAccess struct {
+	LoggingProcessorApacheAccess `yaml:",inline"`
+	LoggingReceiverFilesMixin    `yaml:",inline" validate:"structonly"`
+}
+
+func (r LoggingReceiverApacheAccess) Components(tag string) []fluentbit.Component {
+	if len(r.IncludePaths) == 0 {
+		r.IncludePaths = []string{"/var/log/apache2/access.log"}
```

Comment:
I recall the default paths were different based on VM distros (some examples in https://github.com/GoogleCloudPlatform/ops-agent/blob/master/dev-docs/new-receivers.md#1-research-what-telemetry-the-application-generates)

Is that no longer the case with newer versions, or is it because our integration tests failed to catch it?

Reply:
Ah, looking at the [latest build](https://pantheon.corp.google.com/storage/browser/ops-agents-public-buckets-test-logs/prod/stackdriver_agents/testing/consumer/third_party_apps/presubmit_github/264/20210928-123920/logs), the 3rd party tests only ran on Debian 10. That's probably why the tests did not catch this.

Reply:
I just sent a change internally to extend it to:

```
      'debian-10',
      'centos-8',
      'rhel-7',
      'sles-15',
      'ubuntu-2004-lts',
```

Waiting for review / approval there.

Reply:
This was the main location we found in our research (also in that document it calls out "Apache 2 seems to be dominating. We support that version only"). If desired, we can add the httpd path as well to the default include paths array.

Reply:
+1 that we do not need to support Apache 1.

The issue here is slightly different though. As mentioned in https://github.com/GoogleCloudPlatform/ops-agent/blob/master/dev-docs/new-receivers.md#1-research-what-telemetry-the-application-generates, the access / error log paths differ by distro, even for Apache 2 alone.

```
Default log file paths (RHEL / Red Hat / CentOS / Fedora) | /var/log/httpd/access_log and /var/log/httpd/error_log
Default log file paths (Debian / Ubuntu) | /var/log/apache2/access.log and /var/log/apache2/error.log
```

Spot checked centos 8:
```
[lingshi@centos8 ~]$ sudo service httpd status
Redirecting to /bin/systemctl status httpd.service
● httpd.service - The Apache HTTP Server
   Loaded: loaded (/usr/lib/systemd/system/httpd.service; disabled; vendor preset: disabled)
   Active: active (running) since Tue 2021-09-28 20:27:26 UTC; 29s ago
     Docs: man:httpd.service(8)
 Main PID: 49498 (httpd)
   Status: "Running, listening on: port 80"
    Tasks: 213 (limit: 22244)
   Memory: 24.5M
   CGroup: /system.slice/httpd.service
           ├─49498 /usr/sbin/httpd -DFOREGROUND
           ├─49499 /usr/sbin/httpd -DFOREGROUND
           ├─49500 /usr/sbin/httpd -DFOREGROUND
           ├─49501 /usr/sbin/httpd -DFOREGROUND
           └─49502 /usr/sbin/httpd -DFOREGROUND

Sep 28 20:27:26 centos8 systemd[1]: Starting The Apache HTTP Server...
Sep 28 20:27:26 centos8 systemd[1]: Started The Apache HTTP Server.
Sep 28 20:27:26 centos8 httpd[49498]: Server configured, listening on: port 80
[lingshi@centos8 ~]$ sudo cat /var/log/apache2/access.log
cat: /var/log/apache2/access.log: No such file or directory
[lingshi@centos8 ~]$ sudo cat /var/log/httpd/access_log
[lingshi@centos8 ~]$ httpd -v
Server version: Apache/2.4.37 (centos)
Server built:   May 20 2021 04:33:06
```

Reply:
Okay, I've updated the default include paths to support both common default paths.

Reply:
Looks great. Thanks!